firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationSquasher.java

```diff
@@ -0,0 +1,249 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.model.mutation;
+
+import static com.google.firebase.firestore.util.Assert.hardAssert;
+
+import com.google.firebase.Timestamp;
+import com.google.firebase.firestore.model.DocumentKey;
+import com.google.firebase.firestore.model.FieldPath;
+import com.google.firebase.firestore.model.MutableDocument;
+import com.google.firebase.firestore.model.ObjectValue;
+import com.google.firebase.firestore.model.SnapshotVersion;
+import com.google.firestore.v1.Value;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Squashes multiple mutations applied on a document into a final mutation. It is guaranteed when
+ * the final mutation is applied on a document, it is as if the entire mutation list has been
+ * applied.
+ */
+public class MutationSquasher {
```

Comment:
If we leave Mutations as is, then I don't think we should have this as a separate class. I think that we can use this design if we also combine the Mutation hierarchy into one class. Is that something that you can prototype first? We could then base this PR off this change, which would make it much cleaner.

I will try to find some time to experiment with moving the logic into the individual mutation classes as  well.